### ISSUE TITLE
Memory leak when writing large packets using TCP socket write

### ISSUE THREAD

**Issue by gboswood-hookbang** [IBM-Swift/BlueSocket#160]
I've embedded a fairly thorough comment at the location of the leak below.  I'm currently attempting to switch the TCP part to use SwiftSocket instead of BlueSocket, keeping the same structure.  With SwiftSocket the leak goes away, but it is having its own reliability problems and the API is inferior.  I'd rather get this done with BlueSocket.

The app here is using an electro-mechanical rig to capture thousands of images of objects at fine angular offsets and distances.  In operation it captures one image every one to two seconds.   It currently crashes before reaching a thousand images due to memory starvation.  The memory problem disappears if the call to tcpWrite is commented out, with no other changes.  There shouldn't be a buffering problem since the reader (written in Unity in C#) is keeping up with the images as they are sent (1 to 1), and doesn't send a new image capture request until the prior one has been fully received.  The received images have been observed as being correct and complete.

I would guess that any app sending 2MB data chunks would observe the same leak. I'm suspicious the leak may be occurring in the Darwin layer under BlueSocket since I see no references improperly retained in any of the BlueSocket Swift code.

Update: I've switched the TCP part of this code to use SwiftSocket instead, which required some source edits to handle large messages without problems with EAGAIN (they failed to handle this at all), but that does not leak.  It uses a C language interface to Darwin, while BlueSocket uses a Swift version of the Darwin interface. Maybe something about the different interfaces explains the problem.

Here's the relevant code snipper showing usage with an all-BlueSocket implementation: This runs in its own thread via this call during initialization:

```
networkQueue.async {
            self.manageNetworkConnection();
}
```
```
private func manageNetworkConnection() {
        do {
            let udpSocket = try Socket.create(family: Socket.ProtocolFamily.inet,
                                              type: Socket.SocketType.datagram,
                                              proto: Socket.SocketProtocol.udp)

            while(true) {
                var ipData = Data()
               // Listen for server to announce its IP address via UDP broadcasts
                _ = try udpSocket.listen(forMessage: &ipData, on: 9396);
              
                // Once a UDP broadcast is received, connect via TCP/IP and use that for messaging,
                if let ipString = String(data: ipData, encoding: .utf8) {
                    do {
                        let tcpSocket = try Socket.create();
                        try tcpSocket.connect(to: ipString, port: 9395, timeout: 0)

                        // Don't block on read
                        try tcpSocket.setReadTimeout(value: 10)

                        var pingStart = Date()
                        while(true) {
                            var incomingMessage = Data()
                            try _ = tcpSocket.read(into: &incomingMessage)
                            if(tcpSocket.remoteConnectionClosed) {
                                break;
                            }
                           // Simple ping every two seconds
                            let now = Date()
                            if(now.timeIntervalSince(pingStart) > 2) {
                                pingStart = now
                                let pingData = Data()
                                try tcpSocket.write(from: pingData)
                            }
                           // Check for outgoing messages coming from another thread
                            outgoingMessagesSemaphore.wait()

                            while(outgoingMessages.count > 0) {
                                for _ in 1...outgoingMessages.count {
                                    var message = outgoingMessages.removeFirst();
                                    print("Sending image data, length = " + String(message.count))

                                    let messageCountData = Data(bytes: &message.count,
                                                           count: MemoryLayout.size(ofValue: message.count))
                                    
                                     // *** Not sure if this line leaks, it is small 
                                    try tcpSocket.write(from: messageCountData)

                                    // ********* LEAK HERE **********
                                    // This line is definitely leaking by the size of the message after every call.
                                    // The message an HEIC image from an iPhone typically ranging in size
                                    // between 1MB and 2MB.  If just this line is commented out the leak goes
                                    // away.  No errors are being thrown and every message is successfully
                                    // received and read by the program at the other end of the connection.
                                    try tcpSocket.write(from: message)
                                }
                            }
                            outgoingMessagesSemaphore.signal()

                            // If the nonblocking read returned 0, then there's nothing to read and
                            // this loop does nothing
                            if(incomingMessage.count > 0) {
                                if let message = String(data: incomingMessage, encoding: .utf8) {
                                    handleNetworkMessage(message: message)
                                }
                            }
                        }
                        tcpSocket.close()
                    }
                    catch let error {
                        print(error)
                    }
                }
            }
        }
        catch let error {
            print(error)
        }
    }
```

**Comment by billabt** [comment]
Trying to reproduce but so far have been unable. Still trying.

**Comment by gboswood-hookbang** [comment]
Some time later I found that with SwiftSocket the app leaked as well. It seems that when I used Data's Advanced(by:) function to advance the data buffer to retransmit the remainder of a large packet that didn't fully send (sometimes a message could take several partial sends when working wirelessly) a leak would develop.  

Advanced(by:) makes a new buffer advanced by the amount already sent and it seems not all of the copies generated would be freed.  Why I don't know, maybe it is a Swift bug.  It seemed pretty clear I was releasing all the instances I had.  When I changed away from using Advanced(by:) and instead used Data's getBytes function with a range argument representing the part of the buffer that hadn't successfully been sent in the first TCP transmission, the leak completely disappeared.  I've been able to transfer over 30 thousand photos In my biggest test cases without a leak. 

If BlueSocket is using Advanced(by:) to resubmit the remainder of partially transmitted TCP messages then it might be something to be wary of, unless I was somehow misusing it.  The leak I was observing only seemed to occur during partially successful sends so if someone decides to test this further you'll need to create a situation where the TCP send can't send the whole message in one piece. 

I'm going to close this since it is no longer a problem I can say with certainty exists.  Thank you for taking a look at it and sorry for the inconvenience.
